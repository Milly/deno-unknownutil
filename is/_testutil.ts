// It seems 'IsExact' in deno_std is false positive so use `Equal` in type-challenges
// https://github.com/type-challenges/type-challenges/blob/e77262dba62e9254451f661cb4fe5517ffd1d933/utils/index.d.ts#L7-L9
export type Equal<X, Y> = (<T>() => T extends X ? 1 : 2) extends
  (<T>() => T extends Y ? 1 : 2) ? true : false;

export function stringify(x: unknown): string {
  if (x instanceof Date) return `Date(${x.valueOf()})`;
  if (x instanceof Promise) return "Promise";
  if (typeof x === "function") return x.toString();
  if (typeof x === "bigint") return `${x}n`;
  if (typeof x === "symbol") return x.toString();
  return JSON.stringify(x);
}
