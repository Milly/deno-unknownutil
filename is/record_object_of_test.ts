import { assertEquals } from "@std/assert";
import { assertSnapshot } from "@std/testing/snapshot";
import { assertType } from "@std/testing/types";
import { type Equal, testWithExamples } from "../_testutil.ts";
import { is } from "./mod.ts";
import { isRecordObjectOf } from "./record_object_of.ts";

Deno.test("isRecordObjectOf<T>", async (t) => {
  await t.step("returns properly named function", async (t) => {
    await assertSnapshot(t, isRecordObjectOf(is.Number).name);
    await assertSnapshot(t, isRecordObjectOf((_x): _x is string => false).name);
  });
  await t.step("returns proper type predicate", () => {
    const a: unknown = { a: 0 };
    if (isRecordObjectOf(is.Number)(a)) {
      assertType<Equal<typeof a, Record<PropertyKey, number>>>(true);
    }
  });
  await t.step("returns true on T record", () => {
    assertEquals(isRecordObjectOf(is.Number)({ a: 0 }), true);
    assertEquals(isRecordObjectOf(is.String)({ a: "a" }), true);
    assertEquals(isRecordObjectOf(is.Boolean)({ a: true }), true);
  });
  await t.step("returns false on non T record", () => {
    assertEquals(isRecordObjectOf(is.String)({ a: 0 }), false);
    assertEquals(isRecordObjectOf(is.Number)({ a: "a" }), false);
    assertEquals(isRecordObjectOf(is.String)({ a: true }), false);
  });
  await testWithExamples(
    t,
    isRecordObjectOf((_: unknown): _ is unknown => true),
    {
      excludeExamples: ["record"],
    },
  );
});

Deno.test("isRecordObjectOf<T, K>", async (t) => {
  await t.step("returns properly named function", async (t) => {
    await assertSnapshot(t, isRecordObjectOf(is.Number, is.String).name);
    await assertSnapshot(
      t,
      isRecordObjectOf((_x): _x is string => false, is.String).name,
    );
  });
  await t.step("returns proper type predicate", () => {
    const a: unknown = { a: 0 };
    if (isRecordObjectOf(is.Number, is.String)(a)) {
      assertType<Equal<typeof a, Record<string, number>>>(true);
    }
  });
  await t.step("returns true on T record", () => {
    assertEquals(isRecordObjectOf(is.Number, is.String)({ a: 0 }), true);
    assertEquals(isRecordObjectOf(is.String, is.String)({ a: "a" }), true);
    assertEquals(isRecordObjectOf(is.Boolean, is.String)({ a: true }), true);
  });
  await t.step("returns false on non T record", () => {
    assertEquals(isRecordObjectOf(is.String, is.String)({ a: 0 }), false);
    assertEquals(isRecordObjectOf(is.Number, is.String)({ a: "a" }), false);
    assertEquals(isRecordObjectOf(is.String, is.String)({ a: true }), false);
  });
  await t.step("returns false on non K record", () => {
    assertEquals(isRecordObjectOf(is.Number, is.Number)({ a: 0 }), false);
    assertEquals(isRecordObjectOf(is.String, is.Number)({ a: "a" }), false);
    assertEquals(isRecordObjectOf(is.Boolean, is.Number)({ a: true }), false);
  });
  await testWithExamples(
    t,
    isRecordObjectOf((_: unknown): _ is unknown => true),
    {
      excludeExamples: ["record"],
    },
  );
});
