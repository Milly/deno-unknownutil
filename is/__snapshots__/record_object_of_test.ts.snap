export const snapshot = {};

snapshot[`isRecordObjectOf<T> > returns properly named function 1`] = `"isRecordObjectOf(isNumber, undefined)"`;

snapshot[`isRecordObjectOf<T> > returns properly named function 2`] = `"isRecordObjectOf((anonymous), undefined)"`;

snapshot[`isRecordObjectOf<T, K> > returns properly named function 1`] = `"isRecordObjectOf(isNumber, isString)"`;

snapshot[`isRecordObjectOf<T, K> > returns properly named function 2`] = `"isRecordObjectOf((anonymous), isString)"`;
